build/datagen-steps/metrics.js

```diff
@@ -0,0 +1,425 @@
+    jsonOut[csvArray[i].id] = {};
+    Object.keys(csvArray[i]).forEach((key) => {
+      if (key !== "id") {
+        if (isNumeric(csvArray[i][key])) {
+          jsonOut[csvArray[i].id][key] = Number(csvArray[i][key]);
+        } else {
+          jsonOut[csvArray[i].id][key] = null;
+        }
+      }
+    });
+  }
+  return jsonOut;
+}
+
+/**
+ * Transform CSV array in v2 format to JSON object.
+ *
+ * Format of v2 CSV:
+ *
+ * - Each row represents one year of data for a particular FIPS region
+ * - The columns are always `year, fips, value`
+ *
+ * For example:
+ *
+ * ```csv
+ * year, fips, value;
+ * 2013, 370630001011, 17;
+ * 2014, 370630001011, 17;
+ * 2015, 370630001011, 18;
+ * 2013, 370630001012, 13;
+ * 2014, 370630001012, 14;
+ * 2015, 370630001012, 13;
+ * ```
+ *
+ * The output JSON object is the same as for the v1 importer. It is keyed by the FIPS code, and
+ * contains values in the {y_2012: value, ...} format. For example:
+ *
+ * ```json
+ *  {
+ *    "370630001011": {
+ *      "y_2013": 17,
+ *      "y_2014": 17,
+ *      "y_2015": 18
+ *  },
+ *    "370630001012": {
+ *      ...
+ *    }
+ *  }
+ * ```
+ */
+function v2CsvToJson(csvArray) {
+  const jsonOut = {};
+  csvArray.forEach((row) => {
+    if (!("fips" in row && "year" in row && "value" in row)) {
+      throw new TypeError("Each row must have keys fips, year and value");
+    }
+    if (!row.fips || !row.year) {
+      throw new TypeError("Each row must have a valid fips and year value");
+    }
+    if (!(row.fips in jsonOut)) {
+      jsonOut[row.fips] = {};
+    }
+    jsonOut[row.fips][`y_${row.year}`] = isNumeric(row.value) ? Number(row.value) : null;
+  });
+  return jsonOut;
+}
+
+/**
+ * @param {any} destPath
+ * @param {any} metric
+ * @param {any} json
+ */
+async function writeMetricFile(destPath, metric, json) {
+  const outFile = path.join(destPath, `m${metric.metric}.json`);
+  try {
+    await fs.promises.writeFile(outFile, jsonminify(stringify(json, null, "  ")));
+    console.log(`Wrote ${outFile}`);
+  } catch (err) {
+    console.error(`Error on writing ${outFile}: ${err.message}`);
+  }
+}
+
+/**
+ * Asynchronously checks if given filenames represent files that are readable by the current user.
+ *
+ * Returns a `readabilityStatus` object of the form:
+ *
+ *     {
+ *      [fileName: string]: "readable" | "notReadable"
+ *     }
+ *
+ * Does not throw.
+ *
+ * @param {string[]} fileNames
+ * @returns {Promise<({ [fileName: string]: "readable" | "notReadable" })[]}
+ */
+async function checkIfFilesAreReadable(fileNames) {
+  const readabilityStatusArray = await Promise.all(
+    fileNames.map(async (f) => {
+      try {
+        await fs.promises.access(f, fs.constants.R_OK);
+        return { [f]: "readable" };
+      } catch {
+        return { [f]: "notReadable" };
+      }
+    })
+  );
+  // Merge array of status objects into one object
+  return Object.assign({}, ...readabilityStatusArray);
```

Comment:
Nice, I hadn't seen this pattern for assigning lots of values to an object and like it! Cleaner than creating it up-front and assigning inside the `map`